README.md
![daq (2)](https://user-images.githubusercontent.com/77701711/223956028-f2130450-dd69-4515-ba92-156b1402fdf0.png)

 Flight Vehicle DAQ Code Documentation
 -

I'm writing this in hopes that the code written here is both readible and usable for anyone-- even those without coding or DAQ knowledge-- in lab to use. For those who do have coding experience, my hope is that you will be able improve both this code's functionality and efficiency. I have also provided additional code snippets that can help with repetitive functions. 

~~For the love of god please do not make me have to error check for user inputs~~ <br>
looks like I have to hahaah...fuck

### Config Files
Config files will all be located in the config folder. The file names are hardcoded into the DAQ code and should **NOT** have their name changed unless you know what you are doing. 
#### config.ini
This file currentally has 3 header sections that allow for customization on the server and client side. Note* Omitting a value or submitting an invalid one will result in error:
- [address]
    - ip (str) 
        - see "DAQ Startup Procedure" to find the IP of your system. 
        - An invalid ip will prevent the system from running
    - port (int) 
        - select a port for the socket to bind to. 
        - Must be between 1024 and 49151 otherwise an error is printed
- [server]
    - savefile (str) specifies the name of the file where the data should be stored server side. 
        -Will be in the form: <saveFile><timestamp>.txt in the folder labeled "data"
- [client]
    - pt_poll (float) 
        - Sets the polling rate of the PTs in Hz 
    - sendrate (float)
        - Sets the send rate of the client in Hz

#### PT_Config_FV.ini
This file contains a header for each PT and has the following values:
- port (str)
    - Specifies the port used in the DAQ
- slope (float)
    - Used for voltage to value conversion
- offset (float)
    - Used for voltage to value conversion
- unit (str)
    - Units for PT readings (psi)
- max (int)
    - Max pressure the PT is rated for
- system (str)
    - Specifies which system the PT is configured for (FV)
    
#### TC_Config_FV.ini
This file contains a header for each TC and has the following values:
- port
    - Specifies the port used in the DAQ
- offset
    - Used for voltage to value conversion
- offset_units
    - Specifies the unit of the reading
- temp_units
    - Specifies the unit of the reading
- system (str)
    - Specifies which system the PT is configured for (FV)
    
#### SV_Config_FV.ini
This file contains a header for eaach SV and has the following values:
- port 
    - Specifies the port used in the DAQ
    
#### TC_Port_ID.ini
ID values for TC
- id
    - ID hash for TC validation

### Networking Message (Backend)
These are the possible messages that are sent between the client and server. Users will never explicitly see these messages and only serve as communication between the client and the server. Every message will start with "#" and be divided by "/" 
#### Server side
```
- #KILL 
    Sends a command setting all valves to normal state
- #<valveName>/<proposedChange>
    Sets a valve to a proposed state (ON/OFF)
```
#### Client side
```
- #<sensorName>/<value>/<timestamp>
    Sends a message with the sensorName, value, and timestamp
- #<valveName>/<newChange>
    Sends a message that confirms the proposed change
```
        

Additional Programs
-

Things to Implement:
 - PT, TC, and SV file validity
 - Abort sequence
 - Get custom value timings working
 - Valve timing feedback messages
 - console thread
 - launch sequence
 
Resources
-
 - DAQ Startup Procedure
 - Github Guide for DAQ

Notes
-
For hooking up wires (from left to right) follow this riddle: <br>
    Red touch yellow, kill a fellow,<br>
    Red touch black, won't fry jack, <br>
    Yellow touch red and it'll be dead <br>
Solution: Red (+) Black (-) Yellow (Data)









CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Useful links: 
[Start Up Procedure:](https://docs.google.com/document/d/1HFaLLYg40X4o-dengaRjggNb5vt6SatAqXU8YXTvmBo/edit?usp=sharing)

## [1.1.0] - 2023-3-31
Quick update this time around just in time for the first cold flow on the fuel side (what a coincidence). Lodecell implementation and console access will be in the near future. Special thanks to Galkathe for their contribute to this update.

### What's New
- Stepper motor base implementation
- Value error handling on PT ini file
- Added some descriptive comments to the code
- Spam-proofed the reg buttons (for all you trigger happy folks)
- Updated GUI

### Bug Fixes
- Removed some unnecessary print statements on startup
- Fixed a divide by 0 error if pt_poll or sendrate in config.ini were set to 0 (will set the sleep to 0 instead)
- Autogenerates data directory if it is not in the source code

## [1.0.0] - 2023-03-06
First working version of the DAQ!!! 🥳🥳🥳
Petition to make today a national LPL holiday

### What's New
- Basic DAQ server and client functionality
- PT and TC sensors
- Solenoid actuation (one per command)
- Data logging
- Server kill prototype

verify.py
from configparser import ConfigParser
import os

import timing
#returns a dict of the consoleType members if the file is valid and None if the file does not exist
def verifyExistence(filepath:str, consoleType:str) -> dict:
    parser = ConfigParser()
    try:
        with open(filepath) as f:
            parser.read_file(f)
    except IOError:
        print("ERROR:", filepath, "not found")
        exit()
    #parser = parser.sections()
    if consoleType not in parser.sections():
        print("ERROR: [", consoleType, "] is missing from ", filepath, sep = "")
        print("File rejected for the above reasons. Please fix and restart")
        
        exit()
    fileDict = dict()
    for i in parser.items(consoleType):
        fileDict[i[0]] = i[1]
    return fileDict


#checks file validity and outputs ip and port
def getIPAddress(filepath):
    valid = True
    ipAddress = verifyExistence(filepath, "address")
    
    if "ip" not in ipAddress:
        valid = False
        print("ERROR: ip not specified in [address] section")
    else:
        ip = ipAddress["ip"]
    if "port" not in ipAddress:
        valid = False
        print("ERROR: port is not specifed in the [address] section")
    else:
        try:
            port = int(ipAddress["port"])
            if port < 1024 or port >49151:
                print("ERROR: port ranges can only be from 1024 - 49151")
                valid = False
        except:
            valid = False
            print("ERROR: port is an invalid int")

    if not valid:
        print("File rejected for the above reasons. Please fix and restart")
        exit()
    else:
        return ip, port


#returns dict with necessary server values (savefile for now)
def verifyServerIni(filepath:str):
    parser = verifyExistence(filepath, "server")
    if parser == None:
        exit()
        #assumes the file is valid unless proven otherwise
    valid = True
    serverDict = dict()
    try:
        if "savefile" not in parser:
            valid = False
            print("ERROR: savefile not specified")
        else:
            if not os.path.exists("data"):
                print("WARNING: 'data' directory does not exist. Generating directory...")
                os.makedirs("data")
            serverDict["fp"] = open("data/" + parser["savefile"] + str(timing.missionTime()), 'w')
    except:
        print("ERROR: Invalid file name.")
        valid = False
    serverDict["ip"], serverDict["port"] = getIPAddress(filepath)
    if serverDict["ip"] == None or serverDict["port"] == None:
        valid = False
    if not valid:
        print("File rejected for the above reasons. Please fix and restart")
        exit()
    return serverDict

#client ini file verification
def verifyClientIni(filepath:str):
    parser = verifyExistence(filepath, "client")
    if parser == None:
        exit()
    valid = True
    clientDict = dict()
    if "pt_poll" not in parser:
        valid = False
    else:
        try:
            parsedVal = float(parser["pt_poll"])
            if parsedVal == 0: #divide by 0 error
                clientDict["pt_poll"] = 0
            else: #converts freq to period
                clientDict["pt_poll"] = 1.0 / parsedVal 
        except:
            print("ERROR: [pt_poll] must be a float")
            valid = False
    if "sendrate" not in parser:
        valid = False
    else:
        try:
            parsedVal = float(parser["sendrate"])
            if parsedVal == 0: #divide by 0 error
                clientDict["sendrate"] = 0
            else: #converts freq to period
                clientDict["sendrate"] = 1.0 / parsedVal
        except Exception as e:
            print(e)
            print("ERROR: [sendrate] must be a float")
            valid = False
    if not valid:
        print("File rejected for the above reasons. Please fix and restart")
        exit()
    return clientDict